pkg/sql/submitter.go

```diff
@@ -0,0 +1,240 @@
+}
+
+func submitter() Submitter {
+	s := submitterRegistry[envSubmitter]
+	if s == nil {
+		s = submitterRegistry["default"]
+	}
+	return s
+}
+
+// Submitter extends ir.Executor
+type Submitter interface {
+	ir.Executor
+	Setup(*PipeWriter, *DB, string, *pb.Session) error
+	Teardown()
+	GetTrainIRFromModel() bool
+}
+
+type logChanWriter struct {
+	wr   *PipeWriter
+	m    sync.Mutex
+	buf  bytes.Buffer
+	prev string
+}
+
+func (cw *logChanWriter) Write(p []byte) (n int, err error) {
+	// Both cmd.Stdout and cmd.Stderr are writing to cw
+	cw.m.Lock()
+	defer cw.m.Unlock()
+
+	n, err = cw.buf.Write(p)
+	if err != nil {
+		return n, err
+	}
+	for {
+		line, err := cw.buf.ReadString('\n')
+		cw.prev = cw.prev + line
+		// ReadString returns err != nil if and only if the returned Data
+		// does not end in delim.
+		if err != nil {
+			break
+		}
+		if err := cw.wr.Write(cw.prev); err != nil {
+			return len(cw.prev), err
+		}
+		cw.prev = ""
+	}
+	return n, nil
+}
+
+func (cw *logChanWriter) Close() {
+	if len(cw.prev) > 0 {
+		cw.wr.Write(cw.prev)
+		cw.prev = ""
+	}
+}
+
+type defaultSubmitter struct {
+	Writer   *PipeWriter
+	Db       *DB
+	ModelDir string
+	Cwd      string
+	Session  *pb.Session
+}
+
+type elasticdlSubmitter struct{ *defaultSubmitter }
+type alpsSubmitter struct{ *defaultSubmitter }
+
+func (s *defaultSubmitter) Setup(w *PipeWriter, db *DB, modelDir string, session *pb.Session) error {
+	// cwd is used to store train scripts and save output models.
+	cwd, err := ioutil.TempDir("/tmp", "sqlflow")
+	s.Writer, s.Db, s.ModelDir, s.Cwd, s.Session = w, db, modelDir, cwd, session
+	return err
+}
+
+func (s *defaultSubmitter) SaveModel(cl *ir.TrainClause) error {
+	m := model{workDir: s.Cwd, TrainSelect: cl.OriginalSQL}
+	if s.ModelDir != "" {
+		return m.saveTar(s.ModelDir, cl.Into)
+	}
+	return m.save(s.Db, cl.Into, s.Session)
+}
+
+func (s *defaultSubmitter) LoadModel(cl *ir.TrainClause) error {
+	if s.ModelDir != "" {
+		_, err := loadTar(s.ModelDir, s.Cwd, cl.Into)
+		return err
+	}
+	_, err := load(s.Db, cl.Into, s.Cwd)
+	return err
+}
+
+func (s *defaultSubmitter) runCommand(program string) error {
+	cw := &logChanWriter{wr: s.Writer}
+	var output bytes.Buffer
+	w := io.MultiWriter(cw, &output)
+	defer cw.Close()
+	cmd := sqlflowCmd(s.Cwd, s.Db.driverName)
+	cmd.Stdin, cmd.Stdout, cmd.Stderr = bytes.NewBufferString(program), w, w
+	if e := cmd.Run(); e != nil {
+		return fmt.Errorf("failed: %v\n%sProgram%[2]s\n%s\n%[2]sOutput%[2]s\n%[4]v", e, "==========", program, output.String())
+	}
+	return nil
+}
+
+func (s *defaultSubmitter) ExecuteQuery(sql *ir.StandardSQL) error {
+	return runStandardSQL(s.Writer, string(*sql), s.Db)
+}
+
+func (s *defaultSubmitter) ExecuteTrain(cl *ir.TrainClause) (e error) {
+	var code string
+	if isXGBoostModel(cl.Estimator) {
+		code, e = xgboost.Train(cl)
+	} else {
+		code, e = tensorflow.Train(cl)
+	}
+	if e == nil {
+		if e = s.runCommand(code); e == nil {
+			e = s.SaveModel(cl)
```

Comment:
Style guide: "Try to keep the normal code path at a minimal indentation, and indent the error handling, dealing with it first. This improves the readability of the code by permitting visually scanning the normal path quickly." https://github.com/golang/go/wiki/CodeReviewComments#indent-error-flow

So maybe rewrite this function to following in the next PR. This comment also applies to other places in this PR.
```go
func (s *defaultSubmitter) ExecuteTrain(cl *ir.TrainClause) (e error) {
	var code string
	if isXGBoostModel(cl.Estimator) {
		if code, e = xgboost.Train(cl); e != nil {
			return e
		}
	} else {
		if code, e = tensorflow.Train(cl); e != nil {
			return e
		}
	}
	
	if e = s.runCommand(code); e != nil {
		return e
	}
	
	return s.SaveModel(cl)
}
```

I haven't seen `if e == nil` much in Go project, so please avoid it as much as possible.

Reply:
I'll fix this later in another commit.